### PR TITLE
[WIP] Initial implementation of adversarial training, categorical batch normalization, categorical weight decay, categorical skip connection, categorical multibranch choice

### PR DESCRIPTION
**Pull request by ArlindKadra**
- [x] FGSM Adversarial Training
- [x] Activation controlled BN
- [x] Activation controlled WD
- [x] Activation controlled skip connections, the multi-branch choice is also a hyperparameter now. 
- [x] Add unit tests

### CODE REVIEW

File: autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            typing.Dict[str, np.ndarray]: arguments to the criterion function.
+        """
+        X_adversarial = self.fgsm_attack(X, y)
+        return (X, X_adversarial), {'y_a': y}
+
+    def criterion_preparation(self, y_a: np.ndarray, y_b: np.ndarray = None, lam: float = 1.0
+                              ) -> typing.Callable:
+        # Initial implementation, consider the adversarial loss and the normal network loss
+        # equally.
+        return lambda criterion, pred, adversarial_pred: 0.5 * criterion(pred, y_a) + \
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py:59]: does it need to be fixed? maybe we can sample randomly or set lam: 0.5 in data_preparation

Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py:59]: That is a really good point. I made it fixed since that is how it was in the paper and like that it cares equally about both. 
It makes sense for it to be a hyperparameter too though.

Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py:56]: okay then we can add a hyperparameter later, when we merge this into refactor_development

Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py:56]: Yep, that was exactly what I was thinking.

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
Comment for the function def `get_hyperparameter_search_space` in `autoPyTorch/pipeline/components/setup/network_backbone/ResNetBackbone.py`.

Tuple[Tuple[bool], bool]?
I think if `use_batch_norm=False` then deactivate it and `use_batch_norm=True` then creating config space looks better.
In any case, there are only six possible choices for this.
```
((True, False), True)
((True, False), False)

no need to define config space for the candidates below, because it is a fixed parameter.
((True), True)
((True), False)
((False), True)
((False), False)
```

Based on this, we have four choices:
if we fix the parameter or not and the default parameter,
so should be `use_batch_norm` (default) should be bool and `search_for_batchnorm` (or any other name) should be another bool variable.

File: autoPyTorch/pipeline/components/setup/network_backbone/ShapedResNetBackbone.py [hunk 2]
```diff
@@ -153,8 +154,11 @@ def get_hyperparameter_search_space(dataset_properties: Optional[Dict] = None,
 
         use_dropout = CategoricalHyperparameter(
             "use_dropout", choices=use_dropout[0], default_value=use_dropout[1])
+        use_batch_normalization = CategoricalHyperparameter(
+            "use_batch_norm", choices=use_batch_norm[0], default_value=use_batch_norm[1])
```
[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/network_backbone/ShapedResNetBackbone.py:169]: I did not use the `use_batch_norm` name, because it shadows the variable from the method.

[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/network_backbone/ShapedResNetBackbone.py:169]: > Comment for the function def `get_hyperparameter_search_space` in `autoPyTorch/pipeline/components/setup/network_backbone/ResNetBackbone.py`.
> 
> Tuple[Tuple[bool], bool]?
> I think if `use_batch_norm=False` then deactivate it and `use_batch_norm=True` then creating config space looks better.
> In any case, there are only six possible choices for this.
> 
> ```
> ((True, False), True)
> ((True, False), False)
> 
> no need to define config space for the candidates below, because it is a fixed parameter.
> ((True), True)
> ((True), False)
> ((False), True)
> ((False), False)
> ```
> 
> Based on this, we have four choices:
> if we fix the parameter or not and the default parameter,
> so should be `use_batch_norm` (default) should be bool and `search_for_batchnorm` (or any other name) should be another bool variable.

@nabenabe0928 I disagree, because this would be inconsistent with `use_dropout`. It can be improved in the future.

File: autoPyTorch/pipeline/components/setup/network_backbone/ShapedResNetBackbone.py [hunk 3]
```diff
@@ -82,6 +82,7 @@ def get_properties(dataset_properties: Optional[Dict[str, Any]] = None) -> Dict[
     @staticmethod
     def get_hyperparameter_search_space(dataset_properties: Optional[Dict] = None,  # type: ignore[override]
                                         num_groups: Tuple[Tuple, int] = ((1, 15), 5),
+                                        use_batch_norm: Tuple[Tuple, bool] = ((True, False), True),
```
[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/network_backbone/ShapedResNetBackbone.py:85]: Same reasoning for this one.

**Review by ravinkohli — APPROVED** [review]
